docs/architecture/ErrorReporting.md

```diff
@@ -0,0 +1,122 @@
+# Handling and reporting errors
+
+    WARNING: This document describes future plans for error handling.
+    The UI library does not (yet) implement error handling as described here.
+
+This document describes how errors are handled within the ACS UI library and how they are reported to the surrounding application.
+
+## Error sources
+
+There are two categories of errors, and they are handled differently.
+
+1. ACS backend errors: Errors returned by the REST calls to ACS backend services.
+
+   1. These can be transient or permanent (special case: permission error).
```

Comment:
What is the permission error here? (if the likes of video permission is handled in category 2)

Reply:
Ah I see there's an example below, can you add the example here too

Reply:
The scope here is limited to errors returned from backend REST calls, so video permission errors are covered by (2).

You raise an important point though: How do we report / surface errors in camera permissions? This document currently states that we don't.

* If we don't have permissions: We don't list devices in the device manager / gray out "start video" button etc.
* If we lose those permissions after the call has started (e.g., user went and removed those permissions at the browser level):
  * In the current implementation, the headless SDK handles the actual rendering. It might be that it catches and reports errors in rendering. If that's the case, the error will get handled as in (1) (even though I think that's conceptually wrong). Assuming that the headless SDK doesn't catch the errors...
  * my assumption is that we can't tell the difference between failure modes: "Camera disconnected", "permission revoked", "OS level driver failure" should all look the same at the browser API for security reasons.
  * My recommended approach is to surface the error in the `VideoTile` -- show an error screen for the video that failed to render.
  * Surfacing these errors back to Contoso is challenging: adapter is an API surface for the ACS backend but the video rendering is a purely UI operation.

Let's discuss this problem a bit more. I'd like to add this to the doc in some form.

Reply:
My thinking was adapter isn't an ACS backend API surface, but just the business logic driving the UI (note: it can be totally detached from ACS and we can plug in any custom adapter).
The adapter currently handles the turning on and off of video (despite that not being a backend service driven feature) so I think its OK to surface this error in the adapter onError event.

My initial feeling is this should be a third category of error but not sure what that is as I don't want to be too specific (browser permissions error?)

Reply:
The sources of errors are still classifiable into two: from ACS backend / everything else.

I've reworded the composite section to treat this better. See if that helps.

Reply:
Separated out components and composites entirely. PTAL.